src/lib/viewers/ResultsViewer.ts

```diff
@@ -0,0 +1,136 @@
+		}
+	}
+
+	private countUniqueFiles(violations: Violation[]): number {
+		const fileSet: Set<string> = new Set();
+		violations.forEach(v => {
+			const primaryLocation = v.getCodeLocations()[v.getPrimaryLocationIndex()];
+			const file = primaryLocation.getFile();
+			if (typeof file === 'string') {
+				fileSet.add(file);
+			}
+		});
+		return fileSet.size;
+	}
+
+	private sortViolations(violations: Violation[]): Violation[] {
+		return violations.toSorted((v1, v2) => {
+			// First compare severities.
+			const v1Sev = v1.getRule().getSeverityLevel();
+			const v2Sev = v2.getRule().getSeverityLevel();
+			if (v1Sev !== v2Sev) {
+				return v1Sev < v2Sev ? -1 : 1;
+			}
+			// Next, compare file names.
+			const v1PrimaryLocation = v1.getCodeLocations()[v1.getPrimaryLocationIndex()];
+			const v2PrimaryLocation = v2.getCodeLocations()[v2.getPrimaryLocationIndex()];
+			const v1File = v1PrimaryLocation.getFile() || '';
+			const v2File = v2PrimaryLocation.getFile() || '';
+			if (v1File !== v2File) {
+				return v1File.localeCompare(v2File);
+			}
+
+			// Next, compare start lines.
+			const v1StartLine = v1PrimaryLocation.getStartLine() || 0;
+			const v2StartLine = v2PrimaryLocation.getStartLine() || 0;
+			if (v1StartLine !== v2StartLine) {
+				return v1StartLine - v2StartLine;
+			}
+
+			// Finally, compare start columns.
+			const v1StartColumn = v1PrimaryLocation.getStartColumn() || 0;
+			const v2StartColumn = v2PrimaryLocation.getStartColumn() || 0;
+			return v1StartColumn - v2StartColumn;
+		});
+	}
+
+	private displaySummary(violations: Violation[]): void {
+		const violationCount = violations.length;
+		const fileCount = this.countUniqueFiles(violations);
+		this.display.displayInfo(getMessage(BundleName.ResultsViewer, 'summary.found-results', [violationCount, fileCount]));
+	}
+
+	private displayDetails(violations: Violation[]): void {
+		// If there are more results than the cutoff amount, we'll only show that many.
+		const printableViolationCount = Math.min(ResultsDetailViewer.RESULTS_CUTOFF, violations.length);
+		const omittedResultsCount = violations.length - printableViolationCount;
+		for (let i = 0; i < printableViolationCount; i++) {
+			const violation = violations[i];
+			const rule = violation.getRule();
+			const sev = rule.getSeverityLevel();
+			this.display.displayStyledHeader(
+				getMessage(
+					BundleName.ResultsViewer,
+					'summary.detail.violation-header',
+					[i + 1, rule.getName()])
+			);
+			const primaryLocation = violation.getCodeLocations()[violation.getPrimaryLocationIndex()];
+			this.display.displayStyledObject({
+				engine: rule.getEngineName(),
+				severity: `${sev.valueOf()} (${SeverityLevel[sev]})`,
+				message: violation.getMessage(),
+				location: `${primaryLocation.getFile()}:${primaryLocation.getStartLine()}:${primaryLocation.getStartColumn()}`,
+				docs: violation.getResourceUrls().join(',')
+			}, ['engine', 'severity', 'message', 'location', 'docs']);
```

Comment:
Instead of docs, I think the plan was to call this resources... am I right in that? RetireJs for example gives links to the github PRs/issues for example associated with known vulnerabilities. Also in the future we'll be adding more links that wouldn't be considered "docs".

Reply:
I called it that because it was `docs` in the draft. But changing it to `resources` is trivial if we want to.